src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,518 @@
+        NoSeriesCopilot.Run();
+    end;
+
+    procedure Generate(var NoSeriesProposal: Record "No. Series Proposal"; var ResponseText: Text; var NoSeriesGenerated: Record "No. Series Proposal Line"; InputText: Text)
+    var
+        TokenCountImpl: Codeunit "AOAI Token";
+        SystemPromptTxt: SecretText;
+        CompletePromptTokenCount: Integer;
+        Completion: Text;
+    begin
+        Clear(ResponseText);
+        SystemPromptTxt := GetToolsSystemPrompt();
+
+        CompletePromptTokenCount := TokenCountImpl.GetGPT4TokenCount(SystemPromptTxt) + TokenCountImpl.GetGPT4TokenCount(InputText);
+        if CompletePromptTokenCount <= MaxInputTokens() then begin
+            Completion := GenerateNoSeries(SystemPromptTxt, InputText);
+            if CheckIfValidCompletion(Completion) then begin
+                SaveGenerationHistory(NoSeriesProposal, InputText);
+                CreateNoSeries(NoSeriesProposal, NoSeriesGenerated, Completion);
+            end else
+                ResponseText := Completion;
+        end;
+    end;
+
+    procedure ApplyProposedNoSeries(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    begin
+        if NoSeriesGenerated.FindSet() then
+            repeat
+                InsertNoSeriesWithLines(NoSeriesGenerated);
+                ApplyNoSeriesToSetup(NoSeriesGenerated);
+            until NoSeriesGenerated.Next() = 0;
+    end;
+
+    local procedure InsertNoSeriesWithLines(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    begin
+        InsertNoSeries(NoSeriesGenerated);
+        InsertNoSeriesLine(NoSeriesGenerated);
+    end;
+
+    local procedure InsertNoSeries(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    var
+        NoSeries: Record "No. Series";
+    begin
+        NoSeries.Init();
+        NoSeries.Code := NoSeriesGenerated."Series Code";
+        NoSeries.Description := NoSeriesGenerated.Description;
+        NoSeries."Manual Nos." := true;
+        NoSeries."Default Nos." := true;
+        //TODO: Check if we need to add more fields here, like "Mask", "No. Series Type", "Reverse Sales VAT No. Series" etc.
+        if not NoSeries.Insert(true) then
+            NoSeries.Modify(true);
+    end;
+
+    local procedure InsertNoSeriesLine(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    var
+        NoSeriesLine: Record "No. Series Line";
+    begin
+        NoSeriesLine.Init();
+        NoSeriesLine."Series Code" := NoSeriesGenerated."Series Code";
+        NoSeriesLine."Line No." := GetNoSeriesLineNo(NoSeriesGenerated."Series Code");
+        NoSeriesLine."Starting No." := NoSeriesGenerated."Starting No.";
+        NoSeriesLine."Ending No." := NoSeriesGenerated."Ending No.";
+        NoSeriesLine."Warning No." := NoSeriesGenerated."Warning No.";
+        NoSeriesLine."Increment-by No." := NoSeriesGenerated."Increment-by No.";
+        //TODO: Check if we need to add more fields here, like "Allow Gaps in Nos.", "Sequence Name" etc.
```

Comment:
I think it's important to set the "Allow Gaps" / Implementation field which depends on the purpose of the no. series. 
I think we default to Normal / no gaps because it's the safe choice but we prefer Sequence / allow gaps because it has better perf.